x-pack/plugins/cases/server/client/metrics/alert_details.ts

```diff
@@ -5,33 +5,118 @@
+import { merge } from 'lodash';
+
+import { AlertHostsMetrics, AlertUsersMetrics, CaseMetricsResponse } from '../../../common/api';
+import { createCaseError } from '../../common/error';
+import { AggregationFields, FrequencyResult, UniqueCountResult } from '../../services/alerts/types';
+import { CasesClient } from '../client';
+import { CasesClientArgs } from '../types';
 import { MetricsHandler } from './types';
 
 export class AlertDetails implements MetricsHandler {
-  /**
-   * This boolean protects against the metrics being queried multiple times. The applyMetrics function could be called
-   * once for each feature. All the metrics will be retrieved when the first applyMetrics is called though.
-   */
-  private retrievedMetrics: boolean = false;
+  private requestedFeatures: Set<AggregationFields> = new Set();
+
+  constructor(
+    private readonly caseId: string,
+    private readonly casesClient: CasesClient,
+    private readonly clientArgs: CasesClientArgs
+  ) {}
 
   public getFeatures(): Set<string> {
     return new Set(['alertHosts', 'alertUsers']);
   }
 
+  public setupFeature(feature: string) {
+    switch (feature) {
+      case 'alertHosts':
+        this.requestedFeatures.add(AggregationFields.Hosts);
+        break;
+      case 'alertUsers':
+        this.requestedFeatures.add(AggregationFields.Users);
+        break;
+    }
+  }
+
   public async compute(): Promise<CaseMetricsResponse> {
-    // we already retrieved the metrics so just return them as they are
-    if (this.retrievedMetrics) {
-      return {};
+    const { alertsService, logger } = this.clientArgs;
+
+    try {
+      const alerts = await this.casesClient.attachments.getAllAlertsAttachToCase({
+        caseId: this.caseId,
+      });
+
+      if (alerts.length <= 0 || this.requestedFeatures.size <= 0) {
+        return this.formatResponse();
+      }
+
+      const requestedFeatures = Array.from(this.requestedFeatures);
+
+      const [frequentValues, counts] = await Promise.all([
+        alertsService.getMostFrequentValuesForFields({
+          fields: requestedFeatures,
+          alerts,
+        }),
+        alertsService.countUniqueValuesForFields({
+          fields: requestedFeatures,
+          alerts,
+        }),
+      ]);
+
+      return this.formatResponse({
+        frequentValues,
+        counts,
+      });
+    } catch (error) {
+      throw createCaseError({
+        message: `Failed to retrieve alerts details attached case id: ${this.caseId}: ${error}`,
+        error,
+        logger,
+      });
     }
+  }
+
+  private formatResponse({
+    frequentValues,
+    counts,
+  }: {
+    frequentValues?: FrequencyResult;
+    counts?: UniqueCountResult;
+  } = {}): AlertMetrics {
+    const { hosts: uniqueHosts, users: uniqueUsers } = frequentValues ?? {};
+    const { totalHosts, totalUsers } = counts ?? {};
 
-    this.retrievedMetrics = true;
+    const mergedMetrics: AlertMetrics = Array.from(this.requestedFeatures).reduce(
```

Comment:
Any ideas for making this less gross 😆

Reply:
I think once we add a third value here (and it's coming, we'll need it for oldest alert I think), we should refactor this to use the requested feature enum to know how to get the results from the totals and hosts etc instead of having to destructure them specifically 🤷‍♂️

Reply:
I think we should create a separate small class (inside the same file) for each metric that implements the following interface.

```
interface AlertMetric {
formatResponse: () => Metric
compute: () => Response
}
```

Then inside the `compute` you can do

```
const [frequentValues, counts] = await Promise.all(Object.keys(this.metrics).map((metric) => metric.compute()))

const res = this.metrics.reduce((acc, metric) => ({
...merge(acc, metric.formatResponse())
}), {})
```

where `this.metrics` is

```
private readonly metrics = new Map([['hosts', new HostsMetric], ['users', new UserMetric]])
```

Reply:
Great idea for splitting them into separate classes. I was trying to do a slight optimization here where instead of doing a query to Elasticsearch per alert metric we bundle them into a single aggregation. So maybe instead of them doing a `compute()` call individually they could build their part of the query and they get combined and passed to the alert service. They could also handle formatting the response like you mentioned which solves my gross switch statement 😄

Reply:
I also like the option of having another `AlertMetricHandler` class to delegate this hosts and user specific things, and put all the work together here in the `MetricHandler`, we could have a `format()` accepting the ES response directly, and `getQuery()` methods there.